### DISCUSSION TITLE
Would lit benefit from a react like $$typeof property to protect against potential xss exploits?

### DISCUSSION THREAD

**Discussion by istrugatskiy** [lit/lit#2277] [Ideas]
Currently the lit
``
html`<p>${input}</p>` 
``
function converts this to 
```
{
    _$litType$: 1,
    strings: [
        '<p>',
        '</p>'
    ],
    values: [
        'hello'
    ]
}
```

This is great as it provides a simple object representation of input passed, but can be problematic.
Take this example:
```
import {html, css, LitElement} from 'lit';
import {customElement, property} from 'lit/decorators.js';

const arbitraryString = `{
    "_$litType$": 1,
    "strings": [
        "<img src=\\"test\\" onerror=\\"alert(1)\\">"
    ],
    "values": []
}`

@customElement('my-component')
export class Component extends LitElement {

  @property()
  input = JSON.parse(arbitraryString);

  render() {
    return html`<p>${this.input}</p>`;
  }
}

```
Here lit will interpret the input as a string. This does require a developer to make a mistake. It is not lit's fault. Nevertheless,
would adding a symbol to the object returned by the html function to check whether to render as a string or as an element be a good idea?

Example with modified function:
``
html`<p>${input}</p>` 
``
function converts this to 
```
{
    $$type: Symbol.for('lit.element')
    _$litType$: 1,
    strings: [
        '<p>',
        '</p>'
    ],
    values: [
        'hello'
    ]
}
```

**Comment by justinfagnani** [comment]
Someone could just as easily add the `$$type` property, yeah? `_$litType$` is already serving as our brand check.

**Reply by istrugatskiy** [reply]
Sure, but that property wouldn't have a symbol. JSON doesn't support symbols, and this would prevent what happened here with react: http://danlec.com/blog/xss-via-a-spoofed-react-element.

**Reply by justinfagnani** [reply]
Thanks for the link - I understand the risk now. I think we can mitigate by doing a non-forgeable form JSON check on the template strings array.

**Reply by aomarks** [reply]
This concern was addressed in https://github.com/lit/lit/pull/2307 and was recently released in lit-html 2.1.0
